File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/GeneratorResultDetailsFragment.java
package tabletop.velocic.com.worldforgerpgtools;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.Map;

import tabletop.velocic.com.worldforgerpgtools.GeneratorDeserializer.ResultItem;

public class GeneratorResultDetailsFragment extends android.support.v4.app.Fragment
{
    private static final String ARG_RESULT_ITEM = "result_item";

    private TextView detailTableName;
    private RecyclerView detailList;

    public static GeneratorResultDetailsFragment newInstance(ResultItem resultItem)
    {
        GeneratorResultDetailsFragment fragment = new GeneratorResultDetailsFragment();

        Bundle args = new Bundle();
        args.putParcelable(ARG_RESULT_ITEM, resultItem);
        fragment.setArguments(args);

        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
    {
        View view = inflater.inflate(R.layout.fragment_generated_item_details, container, false);

        detailTableName = view.findViewById(R.id.generated_item_detail_table_name);
        detailList = view.findViewById(R.id.generated_item_detail_list);

        ResultItem detailItem = getArguments().getParcelable(ARG_RESULT_ITEM);
        detailTableName.setText(detailItem.getName());

        detailList.setLayoutManager(new LinearLayoutManager(getActivity()));
        detailList.setAdapter(
            new ResultDetailsAdapter(
                detailItem
            )
        );

        return view;
    }

    private class ResultDetailsViewHolder extends RecyclerView.ViewHolder
    {
        private String detailItemName;
        private String detailItemDescription;
        private TextView descriptionTitle;
        private TextView descriptionContent;

        public ResultDetailsViewHolder(LayoutInflater inflater, ViewGroup parent)
        {
            super(inflater.inflate(R.layout.list_item_generated_item_details, parent, false));

            descriptionTitle = itemView.findViewById(R.id.user_provided_description_title);
            descriptionContent = itemView.findViewById(R.id.user_provided_description_content);
        }

        public void bind(String detailItemName, String detailItemDescription)
        {
            this.detailItemName = detailItemName;
            this.detailItemDescription = detailItemDescription;

            descriptionTitle.setText(this.detailItemName);
            descriptionContent.setText(this.detailItemDescription);
        }
    }

    private class ResultDetailsAdapter extends RecyclerView.Adapter<ResultDetailsViewHolder>
    {
        private ResultItem detailItem;

        public ResultDetailsAdapter(ResultItem detailItem)
        {
            this.detailItem = detailItem;
        }

        @Override
        public ResultDetailsViewHolder onCreateViewHolder(ViewGroup parent, int viewType)
        {
            LayoutInflater inflater = LayoutInflater.from(getActivity());

            return new ResultDetailsViewHolder(inflater, parent);
        }

        @Override
        public void onBindViewHolder(ResultDetailsViewHolder holder, int position)
        {
            Map.Entry<String, String> entry = detailItem.getDetailDataFieldByIndex(position);

            holder.bind(entry.getKey(), entry.getValue());
        }

        @Override
        public int getItemCount()
        {
            return detailItem.getNumDetailDataFields();
        }

        @Override
        public int getItemViewType(int position)
        {
            return R.layout.list_item_generated_item_details;
        }
    }
}

File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/GeneratorResultDetailsFragment.kt
package tabletop.velocic.com.worldforgerpgtools

import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.os.bundleOf
import kotlinx.android.synthetic.main.fragment_generated_item_details.*
import kotlinx.android.synthetic.main.list_item_generated_item_details.view.*
import tabletop.velocic.com.worldforgerpgtools.persistence.ResultItem

class GeneratorResultDetailsFragment : androidx.fragment.app.Fragment()
{
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_generated_item_details, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val detailItem = arguments?.getParcelable(ARG_RESULT_ITEM) as ResultItem?
            ?: throw IllegalStateException("Attempting to display detail information on a generated" +
                " result item, but was provided no item to display.")

        val validatedActivityInstance = activity
            ?: throw IllegalStateException("GeneratorResultDetailsFragment cannot" +
                " exist without a valid FragmentActivity instance.")

        generated_item_detail_table_name.text = detailItem.name
        generated_item_detail_list.layoutManager = LinearLayoutManager(validatedActivityInstance)
        generated_item_detail_list.adapter = ResultDetailsAdapter(validatedActivityInstance, detailItem)
    }

    companion object {
        const val ARG_RESULT_ITEM = "result_item"

        fun newInstance(resultItem: ResultItem) : GeneratorResultDetailsFragment =
            GeneratorResultDetailsFragment().apply {
                arguments = bundleOf(
                    ARG_RESULT_ITEM to resultItem
                )
            }
    }
}

private class ResultDetailsAdapter(
        private val activity: FragmentActivity,
        private val detailItem: ResultItem
) : RecyclerView.Adapter<ResultDetailsViewHolder>()
{
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ResultDetailsViewHolder {
        val view = LayoutInflater.from(activity).inflate(R.layout.list_item_generated_item_details, parent, false)

        return ResultDetailsViewHolder(view)
    }

    override fun onBindViewHolder(holder: ResultDetailsViewHolder, position: Int) {
        val entry = detailItem.detailData[position]

        holder.bind(entry.name, entry.content)
    }

    override fun getItemCount(): Int = detailItem.detailData.size

    override fun getItemViewType(position: Int): Int = R.layout.list_item_generated_item_details
}
private class ResultDetailsViewHolder(
    view: View
) : RecyclerView.ViewHolder(view)
{
    private val descriptionTitle = itemView.user_provided_description_title
    private val descriptionContent = itemView.user_provided_description_content

    fun bind(detailItemName: String, detailItemDescription: String)
    {
        descriptionTitle.text = detailItemName
        descriptionContent.text = detailItemDescription
    }
}
